src/twisted/internet/defer.py

```diff
@@ -53,6 +53,17 @@ class TimeoutError(Exception):
 
 
 
+class NoCurrentResult(BaseException):
+    """
+    This error is raised when attempting to call L{extractCurrentResult} on a
+    L{Deferred} that has not yet called all of its callbacks.
+
+    This error inherits from L{BaseException} rather than L{Exception} to
+    reduce the likelihood of inadvertingly catching this exception as part of
+    an exception handler for other errors.
+    """
```

Comment:
My assumption is that getting a ``NoCurrentResult`` is going to mean that you need to fix your code and isn't something that you're generally going to be catching. I modeled it somewhat after ``GeneratorExit`` in that it's a ``BaseException`` primarily because it's not expecting to get caught. That being said, it is arguably closer in spirit to something like the error that is thrown when you hit a recursion limit or a ``NameError`` or something along that line, which are all inherited from ``Exception``.

Ultimately though, it's really because in my mind, I couldn't think of a case when you would be expecting to catch the ``NoCurrentResult`` error, except that you might catch it by accident. I'm happy to go either way with it though.